### PR TITLE
Adding util::checkpoint

### PR DESCRIPTION
**Pull request by aserio**
This pull request creates a new utility "checkpoint" which simplifies application checkpointing.

### CODE REVIEW

File: hpx/util/checkpoint.hpp [hunk 1]
```diff
@@ -0,0 +1,516 @@
...
+            template <typename... Ts>
+            checkpoint operator()(checkpoint&& c, Ts&&... ts) const
+            {
+                //Create serialization archive from checkpoint data member
+                hpx::serialization::output_archive ar(c.data);
+                //Serialize data
+                int const sequencer[] = {//Trick to expand the variable pack
+                    (ar << ts,
+                        0)...};      //Takes advantage of the comma operator
+                (void) sequencer;    // Suppress unused param. warnings
```
Comment (hkaiser) [inline hpx/util/checkpoint.hpp]: For MSVC you'll have to write this as:
```
int const sequencer[] = {
    0, (ar << ts, 0)...
};
```
this avoids generating a possibly zero-sized array this compiler is complaining about.

File: hpx/util/checkpoint.hpp [hunk 2]
```diff
@@ -0,0 +1,516 @@
...
+    ///          a checkpoint.
+
+    template <typename T, typename... Ts,
+        typename U =
+            typename std::enable_if<!hpx::traits::is_launch_policy<T>::value &&
+                !std::is_same<typename std::decay<T>::type,
+                    checkpoint>::value>::type>
+    hpx::future<checkpoint> save_checkpoint(T&& t, Ts&&... ts)
+    {
+        {
```
Comment (hkaiser) [inline hpx/util/checkpoint.hpp]: What's the rationale for introducing the extra scope?

File: hpx/util/checkpoint.hpp [hunk 3]
```diff
@@ -0,0 +1,516 @@
...
+        ~checkpoint() = default;
+
+        //Other Constructors
+        checkpoint(char* stream, std::size_t count)
+        {
+            for (std::size_t i = 0; i < count; i++)
+            {
+                data.push_back(*stream);
+                stream++;
+            }
```
Comment (sithhell) [inline hpx/util/checkpoint.hpp]: You should resize the vector and then memcpy the data into it here

File: hpx/util/checkpoint.hpp [hunk 4]
```diff
@@ -0,0 +1,516 @@
...
+
+    template <typename T, typename... Ts,
+        typename U =
+            typename std::enable_if<!hpx::traits::is_launch_policy<T>::value &&
+                !std::is_same<typename std::decay<T>::type,
+                    checkpoint>::value>::type>
+    hpx::future<checkpoint> save_checkpoint(T&& t, Ts&&... ts)
+    {
+        {
+            return hpx::dataflow(detail::save_funct_obj(),
```
Comment (sithhell) [inline hpx/util/checkpoint.hpp]: Shouldn't we use the io service pool here?

Comment (hkaiser) [reply hpx/util/checkpoint.hpp]: > Shouldn't we use the io service pool here?

@sithhell what for?

File: hpx/util/checkpoint.hpp (hunk 3 above)
Comment (hkaiser) [reply hpx/util/checkpoint.hpp]: Yes, I agree - the signatures of these should be:

    checkpoint(std::vector<char> const& stream);
    checkpoint(std::vector<char> && stream);

or if you really must,

    template <typename InIter>
    checkpoint(InIter begin, InIter end);

**Comment by biddisco** [comment]
There are quite a lot of commits that only fix inspect errors / formatting / whitespace etc. 
It would be nice if some of these could be squashed/merged into others.

File: hpx/util/checkpoint.hpp (hunk 1 above)
Comment (aserio) [reply hpx/util/checkpoint.hpp]: Will fix!

File: hpx/util/checkpoint.hpp (hunk 2 above)
Comment (aserio) [reply hpx/util/checkpoint.hpp]: Leftover code. Should I remove all of these extra scopes?

File: hpx/util/checkpoint.hpp (hunk 3 above)
Comment (aserio) [reply hpx/util/checkpoint.hpp]: @hkaiser and I are discussing options to remove/replace this construction entirely as exposing pointers in the interface is a weak design.

File: hpx/util/checkpoint.hpp (hunk 4 above)
Comment (aserio) [reply hpx/util/checkpoint.hpp]: ?

File: hpx/util/checkpoint.hpp (hunk 2 above)
Comment (aserio) [reply hpx/util/checkpoint.hpp]: Removed per discussion

**Pull request merged**